src/gotcha_dl.c

```diff
@@ -5,7 +5,137 @@
 #include "elf_ops.h"
 #include <dlfcn.h>
 
-void* _dl_sym(void* handle, const char* name, void* where);
+/**
+ * structure used to pass lookup addr and return library address.
+ */
+struct Addrs {
+    struct link_map *lmap;
+    ElfW(Addr) lookup_addr;
+    int found;
+};
+/**
+ * This is a callback to get headers for each library.
+ * We check if the caller's virtual address is between base address and the virtual addr + memory size.
+ *
+ * @param info, information about headers
+ * @param size, size of headers
+ * @param data, data from caller
+ * @return
+ */
+int lib_header_callback(struct dl_phdr_info * info, size_t size, void * data) {
+    struct Addrs* addrs = data;
+    if (!addrs->found) {
+        char *name = NULL;
+        for (int i = 0; i < info->dlpi_phnum; ++i) {
+            if (info->dlpi_phdr[i].p_type == PT_LOAD) {
+                ElfW(Addr) base_addr = info->dlpi_addr;
+                if (info->dlpi_addr == 0) base_addr = info->dlpi_phdr[i].p_offset;
+                ElfW(Addr) start_addr = base_addr + info->dlpi_phdr[i].p_vaddr;
+                ElfW(Addr) end_addr = start_addr + info->dlpi_phdr[i].p_memsz;
+                if (addrs->lookup_addr >= start_addr && addrs->lookup_addr <= end_addr) {
+                    name = info->dlpi_name;
+                    break;
+                }
+            }
+        }
+        if (name) {
+            struct link_map *current = addrs->lmap;
+            while (current) {
+                if (strcmp(current->l_name, name) == 0) {
```

Comment:
Mentioned this in comments before I started a review, but we should probably do both a name and a base_address comparison to find the match.

Reply:
fixed